File: frontend/src/app/core/services/obra.service.ts
import { HttpClient } from '@angular/common/http';
import { inject, Injectable } from '@angular/core';
import { Obra, ObraRequest } from '../models';
import { Observable } from 'rxjs/internal/Observable';
import { API } from '../constants/api.constants';

@Injectable({
  providedIn: 'root',
})
export class ObraService {
  private readonly http = inject(HttpClient);

  /** Lista TODAS las obras del profesional autenticado, de todos sus comitentes. */
  listarTodas(): Observable<Obra[]> {
    return this.http.get<Obra[]>(API.OBRAS);
  }

  /** Lista las obras de un comitente específico. */
  listarPorComitente(comitenteId: number): Observable<Obra[]> {
    return this.http.get<Obra[]>(API.COMITENTE_OBRAS(comitenteId));
  }

  obtenerPorId(id: number): Observable<Obra> {
    return this.http.get<Obra>(API.OBRA_BY_ID(id));
  }

  /**
   * Crear obra. El endpoint del backend está anidado bajo comitente:
   * POST /api/comitentes/{comitenteId}/obras
   */
  crear(comitenteId: number, request: ObraRequest): Observable<Obra> {
    return this.http.post<Obra>(API.COMITENTE_OBRAS(comitenteId), request);
  }

  /** Actualizar usa el endpoint standalone PUT /api/obras/{id}. */
  actualizar(id: number, request: ObraRequest): Observable<Obra> {
    return this.http.put<Obra>(API.OBRA_BY_ID(id), request);
  }

  eliminar(id: number): Observable<void> {
    return this.http.delete<void>(API.OBRA_BY_ID(id));
  }
}

File: backend/src/main/java/ar/edu/utn/frc/previsar/services/ObraService.java
package ar.edu.utn.frc.previsar.services;

import ar.edu.utn.frc.previsar.dtos.request.ObraRequestDto;
import ar.edu.utn.frc.previsar.dtos.response.ObraResponseDto;

import java.util.List;

/**
 * Operaciones sobre las obras del profesional autenticado.
 *
 * Las obras se acceden a través de los comitentes del profesional.
 * Cualquier operación valida que la obra pertenezca a un comitente
 * del profesional actual.
 */

public interface ObraService {
    /**
     * Lista todas las obras activas del profesional (de cualquier comitente).
     */
    List<ObraResponseDto> listarTodasMisObras();

    /**
     * Lista las obras activas de un comitente específico del profesional.
     */
    List<ObraResponseDto> listarPorComitente(Long comitenteId);

    /**
     * Obtiene una obra por id. Valida propiedad y que no esté eliminada.
     */
    ObraResponseDto obtenerPorId(Long id);

    /**
     * Crea una nueva obra para un comitente específico del profesional.
     */
    ObraResponseDto crear(Long comitenteId, ObraRequestDto request);

    /**
     * Modifica una obra existente del profesional.
     */
    ObraResponseDto actualizar(Long id, ObraRequestDto request);

    /**
     * Soft delete: marca la obra como eliminada.
     */
    void eliminar(Long id);
}
